x-pack/plugins/maps/public/components/metrics_editor/validated_number_input.tsx

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { Component, ChangeEvent, MouseEvent } from 'react';
+import { EuiFieldNumber, EuiFormRow } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+
+interface State {
+  value: number | string;
+  prevValue: number;
+  errorMessage: string;
+  isValid: boolean;
+}
+
+interface Props {
+  value: number;
+  min: number;
+  max: number;
+  onChange: (value: number) => void;
+  label: string;
+}
+
+function isNumberValid(value: number | string, min: number, max: number) {
+  const parsedValue = parseInt(value.toString(), 10);
+
+  if (isNaN(parsedValue)) {
+    return {
+      isValid: false,
+      errorMessage: i18n.translate('xpack.maps.validatedNumberInput.invalidNumberErrorMessage', {
+        defaultMessage: 'Invalid number',
+      }),
+    };
+  }
+
+  const isValid = parsedValue >= min && parsedValue <= max;
+  return {
+    parsedValue,
+    isValid,
+    errorMessage: isValid
+      ? ''
+      : i18n.translate('xpack.maps.validatedNumberInput.invalidClampErrorMessage', {
+          defaultMessage: 'Value must be between {min} and {max}',
+          values: {
+            min,
+            max,
+          },
+        }),
+  };
+}
+
+export class ValidatedNumberInput extends Component<Props, State> {
+  // @ts-expect-error state populated by getDerivedStateFromProps
+  state: State = {};
+
+  static getDerivedStateFromProps(nextProps: Props, prevState: State) {
```

Comment:
I don't think you need to update state value if props value changes. That adds a lot of complexity that is not really needed since in most cases, this input is being used to change the value and you don't have to worry about the value changing outside of the input.